File: service/basis/serverBuilder/src/ServerBuilder.cpp
//******************************************************************
//
// Copyright 2015 Samsung Electronics All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include <memory>
#include <string>
#include <functional>
#include <unordered_map>
#include <vector>

class ResourceAttributes
{
    public:
        /*
         * @brief Value manipulation helper class.
         *          You may need to wrap OCRepresent::AttributeItem class.
         *          Do not expose the class as the part of API.
         */
        class Item
        {
            public:
                /*
                 * @brief Get the value as the type of T
                 *
                 * @throw BadCastException if the type is not correct.
                 */
                template<typename T>
                T get() const;

                /**
                 * @brief Get the value as the type of T
                 *
                 * @return True if the type is correct.
                 */
                template<typename T>
                bool get(T &var) const;

                // ...
                // we need other helper methods for this class.
                // please refer OCRepresent::AttributeItem class.
        };

        /**
         * @brief There will be container in this class, and we need iterator for that.
         * Implement one or just define alias.
         */
        typedef std::unordered_map<std::string, Item>::const_iterator const_iterator;

        /**
         * @brief std::map style get method.
         */
        Item &operator[](const std::string &key);

        /**
         * @brief std::map style get method.
         *
         * @throw InvalidKeyException If no value for the key.
         */
        Item &at(const std::string &key);

        /**
         * @brief std::map style get method.
         *
         * @throw InvalidKeyException If no value for the key.
         */
        const Item &at(const std::string &key) const;

        /**
         * @return True if there is a value for the key.
         */
        bool contains(const std::string &key) const;

        const_iterator cbegin() const;
        const_iterator cend() const;

        size_t getSize() const;
        bool isEmpty() const;

        template<typename T>
        void insert(const std::string &key, const T &value);

        void remove(const std::string &key);

        void clear();
};

class PrimitiveRequest
{
        std::string getResourceUri() const;
};

class PrimitiveResponse
{
    public:
        static PrimitiveResponse DEFAULT;

        static PrimitiveResponse createResponse(const ResourceAttributes &attrs);
};

class PrimitiveServerResource
{
    private:
        PrimitiveServerResource();
    public:
        class Builder
        {
            public:
                Builder(const std::string &uri, const std::string &type,
                        const std::string &interface);

                Builder &setProperties(uint8_t properties);
                Builder &setDiscoverable(bool discoverable);
                Builder &setObservable(bool observable);

                Builder &setAttributes(const ResourceAttributes &attrs);

                PrimitiveServerResource create() const;
        };

        typedef std::function <
        PrimitiveResponse(const PrimitiveRequest &, const ResourceAttributes &) > GetRequestHandler;
        typedef std::function <
        PrimitiveResponse(const PrimitiveRequest &, const ResourceAttributes &) > SetRequestHandler;

        template<typename T>
        void setAttribute(const std::string &key, const T &value);

        template<typename T>
        T getAttribute() const;

        bool hasAttribute(const std::string &key) const;

        const ResourceAttributes &getAttributes() const;
        ResourceAttributes &getAttributes();

        uint8_t getProperties() const;
        bool isObservable() const;
        bool isDiscoverable() const;

        void setGetRequestHandler(GetRequestHandler);
        void setSetRequestHandler(SetRequestHandler);

        void notify();

        std::string getUri() const;
        std::vector<std::string> getTypes() const;
        std::vector<std::string> getInterfaces() const;
};

void example()
{
    PrimitiveServerResource server = PrimitiveServerResource::Builder("uri",    "type",
                                     "interface").setObservable(true).create();

    server.setAttribute("key1", "value");
    server.setAttribute("key2", 0);

    server.setGetRequestHandler(
        [](const PrimitiveRequest &, const ResourceAttributes &) -> PrimitiveResponse
    {
        // If you want to let the default handler handle the request which sends attributes in the server as the response, return DEFAULT.
        return PrimitiveResponse::DEFAULT;

        // or if you want to send back with user defined response
        ResourceAttributes responseAttrs;
        responseAttrs["key"] = 0;

        return PrimitiveResponse::createResponse(responseAttrs);
    });

    // If you want to notify observers of attributes changed.
    server.notify();
}

File: service/resource-encapsulation/examples/android/REServer/src/com/example/sampleserver/ServerBuilder.java
/******************************************************************
 *
 * Copyright 2015 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/

package com.example.sampleserver;

import org.iotivity.ResourceEncapsulation.common.RCSException;
import org.iotivity.ResourceEncapsulation.common.RCSResourceAttributes;
import org.iotivity.ResourceEncapsulation.server.Builder;
import org.iotivity.ResourceEncapsulation.server.RCSResourceObject.*;
import org.iotivity.ResourceEncapsulation.server.RCSGetResponse;
import org.iotivity.ResourceEncapsulation.server.RCSRequest;
import org.iotivity.ResourceEncapsulation.server.RCSResourceObject;
import org.iotivity.ResourceEncapsulation.server.RCSSetResponse;

import android.os.Message;

public class ServerBuilder {

    public static RCSResourceObject           rcsResourceObject;
    private static ServerBuilderDevActivity   serverBuilderDevActivityInstance;
    private static String                     logMessage;
    private static Message                    msg;

    private final GetRequestHandler           getRequestListener;
    private final SetRequestHandler           setRequestListener;
    private final AddAttributeUpdatedListener attributeUpdatedListener;

    // constructor
    public ServerBuilder() {
        serverBuilderDevActivityInstance = ServerBuilderDevActivity
                .getServerBuilderDevActivity();
        getRequestListener = new GetRequestHandler();
        setRequestListener = new SetRequestHandler();
        attributeUpdatedListener = new AddAttributeUpdatedListener();
    }

    // Create Resource
    public void createResource() {

        Builder builderObject = new Builder("/a/TempSensor",
                "core.TemperatureSensor", "oic.if.");
        builderObject.setDiscoverable(true);
        builderObject.setObservable(true);

        RCSResourceAttributes attr = new RCSResourceAttributes();
        try {
            attr.setValueInt("Temperature", 10);
        } catch (RCSException e) {
            e.printStackTrace();
        }
        builderObject.setAttributes(attr);

        try {
            rcsResourceObject = builderObject.build();
            rcsResourceObject
                    .setSetRequestHandlerPolicy(SetRequestHandlerPolicy.ACCEPTANCE);
        } catch (RCSException e) {
            e.printStackTrace();
        }
    }

    // Set Attribute
    public void setAttribute(int value) {

        rcsResourceObject.setAttribute("Temperature", value);
        logMessage = "Attribute set successfully" + "\n";
        logMessage += "Temperature : " + value;

        serverBuilderDevActivityInstance.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 0;
        serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
    }

    // Set Get Request Listener
    public void setGetRequestListener() {
        rcsResourceObject.setGetRequestListener(getRequestListener);
        rcsResourceObject.setGetRequestHandler();
        logMessage = "Get Handler set successfully." + "\n";

        serverBuilderDevActivityInstance.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 0;
        serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
    }

    // Set Set Request Listener
    public void setSetRequestListener() {
        rcsResourceObject.setSetRequestListener(setRequestListener);
        rcsResourceObject.setSetRequestHandler();
        logMessage = "Set Handler set successfully." + "\n";

        serverBuilderDevActivityInstance.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 0;
        serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
    }

    // Set Attribute Updated Listener
    public void setAttributeUpdatedListener() {

        rcsResourceObject.setAttributeUpdatedListener(attributeUpdatedListener);
        rcsResourceObject.addAttributeUpdatedListener("Temperature");
        logMessage = "Attribte updated listener set successfully." + "\n";

        serverBuilderDevActivityInstance.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 0;
        serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
    }

    // Remove Attribute Updated Listener
    public void removeAttributeUpdatedListener() {

        rcsResourceObject.removeAttributeUpdatedListener("Temperature");
        logMessage = "Attribte updated listener removed successfully." + "\n";

        serverBuilderDevActivityInstance.setMessageLog(logMessage);
        msg = Message.obtain();
        msg.what = 0;
        serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
    }

    /**
     * Listener for receiving Get Request and for Sending Response
     */
    private class GetRequestHandler implements IGetRequestHandler {

        @Override
        public RCSGetResponse onGetRequest(RCSRequest request,
                RCSResourceAttributes attributes) {

            logMessage = "Got a Get request from client sending dafault response \n"
                    + "URI : " + request.getResourceUri() + "\n";
            serverBuilderDevActivityInstance.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 0;
            serverBuilderDevActivityInstance.getHandler().sendMessage(msg);

            RCSGetResponse object = RCSGetResponse.defaultAction();

            return object;
        }
    };

    /**
     * Listener for receiving Set Request and for Sending Response
     */
    private class SetRequestHandler implements ISetRequestHandler {

        @Override
        public RCSSetResponse onSetRequest(RCSRequest request,
                RCSResourceAttributes attributes) {

            logMessage = "Got a Set request from client sending dafault response"
                    + "\n" + "URI : " + request.getResourceUri() + "\n";
            serverBuilderDevActivityInstance.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 0;
            serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
            RCSSetResponse object = RCSSetResponse.defaultAction();
            return object;
        }
    };

    /**
     * Listener for receiving Updated Attribute
     */
    private class AddAttributeUpdatedListener implements
            IAttributeUpdatedListener {
        @Override
        public void onAttributeUpdate(String oldValue, String newValue) {

            logMessage = "attributed updated" + "\n";
            logMessage += "oldValue : " + oldValue + "\n";
            logMessage += "newValue : " + newValue + "\n";
            serverBuilderDevActivityInstance.setMessageLog(logMessage);
            msg = Message.obtain();
            msg.what = 0;
            serverBuilderDevActivityInstance.getHandler().sendMessage(msg);
        }
    };
}
